crates/bitwarden-vault/src/cipher/cipher_risk_client.rs

```diff
@@ -0,0 +1,778 @@
+                    }
+                } else {
+                    ExposedPasswordResult::NotChecked
+                };
+
+                // Check reuse from provided map
+                let reuse_count = if let Some(map) = &password_map {
+                    map.map.get(&details.password).copied()
+                } else {
+                    None
+                };
+
+                CipherRisk {
+                    id: details.id,
+                    password_strength,
+                    exposed_result,
+                    reuse_count,
+                }
+            }
+        });
+
+        // Process up to MAX_CONCURRENT_REQUESTS futures concurrently
+        // Individual HIBP errors are captured per-cipher, so we use collect() instead of
+        // try_collect()
+        let results: Vec<CipherRisk> = stream::iter(futures)
+            .buffer_unordered(MAX_CONCURRENT_REQUESTS)
+            .collect()
+            .await;
+
+        Ok(results)
+    }
+
+    /// Calculate password strength with cipher-specific context.
+    ///
+    /// Uses zxcvbn to score password strength from 0 (weakest) to 4 (strongest).
+    /// Penalizes passwords that contain parts of the username/email.
+    fn calculate_password_strength(password: &str, username: Option<&str>) -> u8 {
+        let mut user_inputs = Vec::new();
+
+        // Extract meaningful parts from username field
+        if let Some(username) = username {
+            user_inputs.extend(Self::extract_user_inputs(username));
+        }
+
+        // Call zxcvbn with cipher-specific inputs only (no "bitwarden" globals)
+        let inputs_refs: Vec<&str> = user_inputs.iter().map(|s| s.as_str()).collect();
+        zxcvbn::zxcvbn(password, &inputs_refs).score().into()
+    }
+
+    /// Extract meaningful tokens from username/email for password penalization.
+    ///
+    /// Handles both email addresses and plain usernames:
+    /// - For emails: extracts and tokenizes the local part (before @)
+    /// - For usernames: tokenizes the entire string
+    /// - Splits on non-alphanumeric characters and converts to lowercase
+    fn extract_user_inputs(username: &str) -> Vec<String> {
+        // Check if it's email-like (contains @)
+        if let Some((local_part, _domain)) = username.split_once('@') {
+            // Email: extract local part tokens
+            local_part
+                .trim()
+                .to_lowercase()
+                .split(|c: char| !c.is_alphanumeric())
+                .filter(|s| !s.is_empty())
+                .map(str::to_owned)
+                .collect()
+        } else {
+            // Username: split on non-alphanumeric
+            username
+                .trim()
+                .to_lowercase()
+                .split(|c: char| !c.is_alphanumeric())
+                .filter(|s| !s.is_empty())
+                .map(str::to_owned)
+                .collect()
+        }
```

Comment:
⛏️ You could use Option::map_or() here to avoid duplicating the string processing chain. 

```suggestion
        username
            // Check if it's email-like (contains @)
            .split_once('@')
            // Email: extract local part tokens
            .map_or(username, |(local_part, _domain)| local_part)
            .trim()
            .to_lowercase()
            .split(|c: char| !c.is_alphanumeric())
            .filter(|s| !s.is_empty())
            .map(str::to_owned)
            .collect()
```

Reply:
👍 [1ea9850](https://github.com/bitwarden/sdk-internal/pull/499/commits/1ea98508fd1069a84e047325bab6f67e6c8c2ef5)